arch/arm/core/aarch32/mpu/arm_core_mpu.c

```diff
@@ -76,7 +76,7 @@ static const struct z_arm_mpu_partition static_regions[] = {
 		/* Special non-cacheable RAM area */
 		.start = (uint32_t)&_nocache_ram_start,
 		.size = (uint32_t)&_nocache_ram_size,
-		.attr = K_MEM_PARTITION_P_RW_U_NA_NOCACHE,
+		.attr = K_MEM_PARTITION_P_RW_U_RW_NOCACHE,
```

Comment:
@ioannisg, @aurel32  Would you mind giving your thoughts on this ?
Change shows that a user space accessible non cacheable partition would make sense for M7 userspace DMA users (simplify use of DMA base transaction while keeping cache related performances, as for now the alternative is to fully disable cache)
So, would it be acceptable to change attributes of this partition ?
If not, would it make sense to add a new partition that would be userspace RW ?

Reply:
> @ioannisg, @aurel32 Would you mind giving your thoughts on this ?
> Change shows that a user space accessible non cacheable partition would make sense for M7 userspace DMA users (simplify use of DMA base transaction while keeping cache related performances, as for now the alternative is to fully disable cache)
> So, would it be acceptable to change attributes of this partition ?

When we discussed adding the NO_CACHE memory, we had in mind that it's only required at the driver levels, that's why we only make it accessible to kernel space. I don't think we can change that, it basically gives access to memory areas that the user space is not supposed to get access to (network data for example), so this is a security issue.

> If not, would it make sense to add a new partition that would be userspace RW ?

That's an option, to consider.

The other alternative is to use cache management operation at the driver level. This is already implemented in some drivers (Atmel Ethernet driver, Atmel and STM32 I2S drivers), and I guess the same can be done for the serial driver. Compare to uncached memory, that should bring some performance improvement, and simpler userland code, in the sense that the memory only need to be declared aligned, not anymore in the nocache area. It's just a pitty we still do not have functions for that in Zephyr and that we have to rely on the CMSIS functions.

Reply:
We clearly lack cache related functions. I guess the same way we have created the `__nocache__` macro, we can create something like a `__dma_alloc__` macro that takes care of aligning the data correctly. This should not be different from the user point of view.

Adding a new MPU region for uncached memory is not trivial either, at least in the `CONFIG_USERSPACE=y`, because you need to swap that MPU region on task change. It also mean we need to create a new __nocache_user__ macro to put the variables in the right sections.

The other short term alternatives is to disable the D-Cache (but not the I-Cache) when `CONFIG_UART_ASYNC_API=y`. 

An other STM32 specific short term alternative is to create `__dma_alloc__` macro that allocates the variables in the DTCM. In the longer term we can imagine to let users choose between allocating those variables in the DTCM or in standard memory with the use of cache management operations.

Reply:
Some additional thinking:
- `__nocache` prevents 1) variable allocated on the stack (the section is silently dropped) 2) initialized. This explains why the uart_async_api test needs to be heavily modified.
- Cache management operations requires that 1) buffer are aligned on a cache line (32-byte) 2) There is proper padding so that no other unrelated variables are placed on the same cache line. I *think* those requirements are actually only true for RX buffer, as a D-Cache clean operation doesn't have side effects besides performance.

Both cases definitely have an impact on the user code, although in a different way.

Reply:
> The other short term alternatives is to disable the D-Cache (but not the I-Cache) when CONFIG_UART_ASYNC_API=y.

That's the easiest indeed. But we need to do it also for other DMA users like SPI, I2S and upcoming ones (hopefully ADC).
Eventually, only some happy few will benefit from full M7 cache performances :-).

> An other STM32 specific short term alternative is to create __dma_alloc__ macro that allocates the variables in the DTCM. 

That's indeed an interesting direction as well. Ideally we can extend that to other M7 + DTCM based architectures.

Thanks @aurel32 for your advised feedback. Let's see what @ioannisg think about that.

Reply:
Defining a contiguous USER NOCACHE memory area to store (`__section__(".usernocache")`) user buffers is possible for this test :
```
                0x0000000020020000                . = 0x20020000
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                _image_ram_start = .

nocache         0x0000000020020000     0x1000
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                . = ALIGN ((0x1 << LOG2CEIL (_nocache_ram_size)))
                0x0000000020020000                _nocache_ram_start = .
 *(SORT_BY_ALIGNMENT(.nocache))
 *(SORT_BY_ALIGNMENT(.nocache.*))
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                . = ALIGN ((0x1 << LOG2CEIL (_nocache_ram_size)))
                0x0000000020020000                _nocache_ram_end = .
                0x0000000020020000                . = ALIGN (_region_min_align)
                0x0000000020020000                . = ALIGN ((0x1 << LOG2CEIL (_nocache_ram_size)))
                0x0000000020020000                _usernocache_ram_start = .
 *(SORT_BY_ALIGNMENT(.usernocache))
 .usernocache   0x0000000020020000      0xf10 app/libapp.a(test_uart_async.c.obj)
 *(SORT_BY_ALIGNMENT(.usernocache.*))
                0x0000000020020f20                . = ALIGN (_region_min_align)
 *fill*         0x0000000020020f10       0x10 
                0x0000000020021000                . = ALIGN ((0x1 << LOG2CEIL (_usernocache_ram_size)))
 *fill*         0x0000000020020f20       0xe0 
                0x0000000020021000                _usernocache_ram_end = .
                0x0000000000000000                _nocache_ram_size = (_nocache_ram_end - _nocache_ram_start)
                0x0000000020020000                _usernocache_ram_start = _nocache_ram_end
                0x0000000000001000                _usernocache_ram_size = (_usernocache_ram_end - _usernocache_ram_start)
```